JBPM-8403: Stunner - Missing adhoc activation condition in adhoc subprocess

wmedvede opened the pull request.

wmedvede commented:
Hello @romartin @hasys , would you mind review this PR?
Thanks

romartin commented:
Jenkins execute full downstream build

wmedvede commented:
full build failed in non related stuff as usual, but artifacts were generated.

03:08:12 [ERROR] Failed to execute goal org.codehaus.cargo:cargo-maven2-plugin:1.7.1:start (start-container) on project business-central-tests-gui: Cannot start container [org.codehaus.cargo.container.wildfly.WildFly14xInstalledLocalContainer@5e9c7360]: Monitor [org.codehaus.cargo.container.spi.startup.CombinedContainerMonitor] failed to detect running container within the timeout period [120000]. -> [Help 1]
03:08:12 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.cargo:cargo-maven2-plugin:1.7.1:start (start-container) on project business-central-tests-gui: Cannot start container [org.codehaus.cargo.container.wildfly.WildFly14xInstalledLocalContainer@5e9c7360]
03:08:12     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
03:08:12     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
03:08:12     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
03:08:12     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
03:

wmedvede commented:
full build failed as usual in non related stuff, but artifacts were generated.

wmedvede commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/properties/AdHocSubProcessPropertyWriter.java:
```diff
@@ -36,6 +38,12 @@ public AdHocSubProcessPropertyWriter(AdHocSubProcess process, VariableScope vari
     }
 
+    public void setAdHocActivationCondition(BaseAdHocActivationCondition adHocActivationCondition) {
+        if (StringUtils.nonEmpty(adHocActivationCondition.getValue())) {
```
Hi @tiagodolphine , thanks for asking. 
I think this is correct, if the value of the adHocActivationCondition property is null or the empty String, it means that there are no ActivationCondition, and in this case we don't want/need to generate an empty or null element in the resulting xml tag. Old jBPM designer works the same way, and it makes sense to me.

Regarding the null check of the adHocActivationCondition itself I believe that with the existing pattern it's not necessary.  I general in the models we have that Definitions are composed of properties (@Property and @PropertySet), and finally the property holds the value. And all the models and programming we have now relies on the fact that when a Definition is created, all the properties/property sets are initialized. So the property can have empty, null, or whatever thing as a value, but the property can not be null by definition.

For example in general all the XXXPropertyWriters do things like this:

        BPMNGeneralSet general = definition.getGeneral();
        p.setName(general.getName().getValue());  //see here, we basically gets the "name" property and asks for his value, etc.
        p.setDocumentation(general.getDocumentation().getValue());

I don't know to be honest the origin of this pattern of initializing all the object graph for a Definition, when in the end at the moment of creating the bean you really don't know how many properties will really have a value. But it can probably be related with the forms, that probably expects the beans are initialized in this way, etc.

wmedvede commented:
full-bld failed, with no information, but artifacts where generated.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/AdHocActivationCondition.java:
```diff
@@ -0,0 +1,75 @@
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
```
Can be:

```java
if (o instanceof AdHocActivationCondition) {
        AdHocActivationCondition that = (AdHocActivationCondition) o;
        return Objects.equals(value, that.value);
}

return false;
```

`instanceof` check for null as well.

hasys commented:
Hi @wmedvede, I am not able to test build at runtime. I downloaded binary yesterday and now it is not possible to log in to Business Central. I see infinite loading:

![image](https://user-images.githubusercontent.com/1477262/61526638-0f5b6e80-aa1b-11e9-929c-0bbc0ecce6b3.png)

I am not able to download it again, because it is already not available on Jenkins. My suggestion is fix merge conflicts and my suggestions and rerun Full Downstream Build. WDYT?

wmedvede commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/processes/SubProcessConverterTest.java:
```diff
@@ -103,4 +127,79 @@ public void testConvertAdHocSubprocessNode_notautostart() {
+        assertEquals(parentY1 + nodeY1, shape.getBounds().getY(), 0);
+        assertEquals(nodeX2 - nodeX1, shape.getBounds().getWidth(), 0);
+        assertEquals(nodeY2 - nodeY1, shape.getBounds().getHeight(), 0);
+        int i = 0;
```
left by mistake, removed, thanks

wmedvede commented:
> Hi @wmedvede, I am not able to test build at runtime. I downloaded binary yesterday and now it is not possible to log in to Business Central. I see infinite loading:
> 
> ![image](https://user-images.githubusercontent.com/1477262/61526638-0f5b6e80-aa1b-11e9-929c-0bbc0ecce6b3.png)
> 
> I am not able to download it again, because it is already not available on Jenkins. My suggestion is fix merge conflicts and my suggestions and rerun Full Downstream Build. WDYT?

yes, I agree. Suggestions applied and rebased with master. Build might still fail until this non related PR is merged https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/998

Thanks

wmedvede commented:
local build failed with the errors below as expected. 
let's wait for https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/998 to be merged

13:47:28 [WARNING] The requested profile "wildfly" could not be activated because it does not exist.
13:47:28 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-blacklisted-dependencies) on project kie-wb-common-compiler-distribution: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
13:47:28 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-blacklisted-dependencies) on project kie-wb-common-compiler-distribution: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.
13:47:28     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
13:47:28     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)

xiezhang7 commented:
> Hello @xiezhang7, @LuboTerifaj, can I ask you to check is sync with CM version of AdHoc sub-process is correct? Is it possible to override BPMN version and remove duplications?

Hi @hasys , it looks good to me. 
I don't think some of the annotations will still work correctly if the duplication of `AdHocActivationCondition` is removed.

wmedvede commented:
jenkins retest this

hasys commented:
Jenkins execute full downstream build

hasys commented:
> > Hello @xiezhang7, @LuboTerifaj, can I ask you to check is sync with CM version of AdHoc sub-process is correct? Is it possible to override BPMN version and remove duplications?
> 
> Hi @hasys , it looks good to me.
> I don't think some of the annotations will still work correctly if the duplication of `AdHocActivationCondition` is removed.

Is it possible to extend process and set-up new annotations? AFAIK annotations are not inherited until they are marked as `@Inherited`, and the arn't. I guess it is quite safe to extend existing classes and mark it by necessary annotations.

hasys commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

xiezhang7 commented:
> Is it possible to extend process and set-up new annotations? AFAIK annotations are not inherited until they are marked as `@Inherited`, and the arn't. I guess it is quite safe to extend existing classes and mark it by necessary annotations.

@hasys  One problem is the legacy marshaller cannot handle inherited properties in some cases. That's why all the properties are duplicated.
Another problem is the label and description of [AdHocActivationCondition](https://github.com/kiegroup/kie-wb-common/pull/2748/files#diff-91c6d0d1d0636920d711716960d69510) are different for case view. Without duplication, the label and description cannot be overridden for case view.

wmedvede commented:
jenkins execute full downstream build

wmedvede commented:
Hi @hasys full-bld faild in non related tests but artifacts were generated.

LuboTerifaj commented:
@hasys I didn't find any issues in CM editor.

wmedvede commented:
Hi @hasys , so after Yesterday's bluejeans we could see that the activation condition is generated well in the bpmn file and works. And I've just added the text help and the (drools) to the AdhocActivationConditon field. It now looks like this.

![Adhoc-Activation-Condition](https://user-images.githubusercontent.com/2431454/61934101-d0cc3380-af87-11e9-92d1-2720038e7411.png)

So I believe this is ready. This was also rebased with last master commits.

wmedvede commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

wmedvede commented:
Hi @hasys help text was changed to "Drools expression that determines when the ad-hoc sub-process will be activated" as suggested.

wmedvede commented:
Jenkins execute full downstream build

wmedvede commented:
@hasys @romartin full-bld failed in non but artifacts where generated.
This last new-build really only include the change in the help text mentioned above. (it's also rebased with master)

I think this is ready to be merged.
Thanks

romartin commented:
yep, missing review from @hasys , thanks guys

wmedvede commented:
Hi @tiagodolphine would you mind approve this PR?
Thanks!

wmedvede commented:
> Hi @wmedvede LGTM.
> Just one last comment, it seems the property label "AdHocActivationCondition" does not have the i18n set.
> Thanks !

Hi @tiagodolphine the i18n actually set to show "AdHocActivationCondition (drools)" and is working fine. And currently we are using the same names as in jBPM Designer, this is why it says AdHocActivationCondition.

tiagodolphine commented:
Ok @wmedvede thanks for the clarification. So, all good to me.

Thanks.
